### DISCUSSION TITLE
Allow `IterableIterator` in methods that can work with arrays (persist, remove, etc)

### DISCUSSION THREAD

**Discussion by ruscon** [mikro-orm/mikro-orm#4723] [Ideas]
**Why is this necessary?**
Make it easier to write code when working with `Map`, `Set` and others

Right now
```ts
persist<Entity extends object>(entity: Entity | Reference<Entity> | (Entity | Reference<Entity>)[]): this;
```

Suggestion
```ts
persist<Entity extends object>((entity: Entity | Reference<Entity>) | IterableIterator<Entity | Reference<Entity>> | (Entity | Reference<Entity>)[]): this;
```

Guard
```ts
export function isIterable<T>(value: unknown): value is Iterable<T> {
  return (
    typeof value === 'object' &&
    value !== null &&
    Symbol.iterator in value &&
    typeof value[Symbol.iterator] === 'function'
  )
}
```

Usage
```ts
persist<Entity extends object>(entity: Entity | Reference<Entity> | IterableIterator<Entity | Reference<Entity>> | (Entity | Reference<Entity>)[]): this {
  entity = isIterable<Entity>(entity) ? [...entity] : entity

  // ... current implementation
}
```

Same for other methods

**Comment by B4nan** [comment]
`Set` is actually already supported, as most (if not all) of those methods that accept arrays as well will use `Utils.asArray()` under the hood:

https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/utils/Utils.ts#L406

I don't think we should support `Map` out of box, as its not clear whether you want the keys (which can be objects as well) or the values - it's up to you to use `map.values()` or `map.keys()` explicitly.

**Reply by ruscon** [reply]
@B4nan No no no...
We are talking specifically about `IterableIterator`

```ts
await em.persist(someMap.values()) // `IterableIterator` is passed 
```
instead of 
```ts
await em.persist([...someMap.values()])
```

if you try to pass the `Map` instance itself you will see the error something like this:
```
TS2345: Argument of type  Map  is not assignable to parameter of type
MyEntity | Reference<MyEntity> | IterableIterator<MyEntity | Reference<MyEntity>> | (MyEntity | Reference<...>)[]
Type  Map<number, MyEntity>  is missing the following properties from type  (MyEntity | Reference<MyEntity>)[] :  length, pop, push, concat , and  24  more.
```

**Reply by B4nan** [reply]
As a side effect, #4798 will make the `Utils.asArray()` method work with iterators too, so now it will be really just about the types - I've added the support to the `Collection` methods and `persist/remove`, feel free to PR other places that are accepting arrays (and already work with `Utils.asArray()` under the hood).
